server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -412,11 +417,68 @@ public void executeQueryPhase(ShardSearchRequest request, SearchShardTask task,
                     return;
                 }
             }
-            // fork the execution in the search thread pool
-            runAsync(getExecutor(shard), () -> executeQueryPhase(orig, task), l);
+            ensureAfterSeqNoRefreshed(shard, orig, () -> executeQueryPhase(orig, task), l);
```

Comment:
We should also check the need for refresh in the can match phase. See https://github.com/elastic/elasticsearch/blob/2791257f62b37679c9dba7e24a447a0eb534f181/server/src/main/java/org/elasticsearch/search/SearchService.java#L1319 where we consider that the shard matches because a refresh is pending (search idle case).

Reply:
I made a change here. Make sure that it is what you had in mind as this code is less familiar to me.